### PR TITLE
Expose theta and phi builders publicly

### PR DESCRIPTION
**Pull request by AdamGleave**
For my use case it's helpful to be able to build the networks outside of the reward shaping classes when desired. I expect this to also be helpful to other users, who might be working outside of `reward_net.py`  but still want to use similar architectures.

I've also generalized `build_basic_theta_network` to be able to depend on the next observation.

*Note*: the `docstrings` base will need to be changed.

### CODE REVIEW

File: imitation/reward_net.py [hunk 1]
```diff
@@ -183,19 +183,42 @@ def build_summaries(self):
...
+  Conversely, if `new_obs_input` and `act_input` are both set to `None`, it
+  will depend just on the current observation: :math:`R(o)`.
+
+  Arguments:
+    hid_sizes (Optional[List[int]]): Number of units at each hidden layer.
+    obs_space (gym.Space): Observation space.
+    act_space (gym.Space): Action space.
+    old_obs_input (Optional[tf.Tensor]): Previous observation.
+    new_obs_input (Optional[tf.Tensor]): Next observation.
+    act_input (tf.Tensor): Action.
```
Comment (shwang) [inline imitation/reward_net.py]: `act_input` should also be `Optional`.

File: imitation/reward_net.py [hunk 2]
```diff
@@ -183,19 +183,42 @@ def build_summaries(self):
...
+                              old_obs_input, new_obs_input, act_input):
+  """Builds a reward network depending on specified observations and actions.
+
+  All specified inputs will be preprocessed and then concatenated. If all
+  inputs are specified, then it will be a :math:`R(o,a,o')` network.
+  Conversely, if `new_obs_input` and `act_input` are both set to `None`, it
+  will depend just on the current observation: :math:`R(o)`.
+
+  Arguments:
+    hid_sizes (Optional[List[int]]): Number of units at each hidden layer.
```
Comment (shwang) [inline imitation/reward_net.py]: Note that default is linear

File: imitation/reward_net.py [hunk 3]
```diff
@@ -224,16 +247,50 @@ def __init__(self, env, *, state_only=False, theta_units=None, **kwargs):
...
   def reward_output_train(self):
     """Training reward is the same as the test reward, since no shaping."""
     return self.reward_output_test
 
 
+def build_basic_phi_network(hid_sizes, obs_space, old_obs_input, new_obs_input):
+  """Builds a potential network depending on specified observation.
+
+  Arguments:
+    hid_sizes (Optional[List[int]]): Number of units at each hidden layer.
```
Comment (shwang) [inline imitation/reward_net.py]: Note that default is [32, 32]

File: imitation/reward_net.py [hunk 4]
```diff
@@ -224,16 +247,50 @@ def __init__(self, env, *, state_only=False, theta_units=None, **kwargs):
...
+    obs_space (gym.Space): Observation space.
+    old_obs_input (Optional[tf.Tensor]): Previous observation.
+    new_obs_input (Optional[tf.Tensor]): Next observation.
+
+  Returns:
+    Tuple[tf.Tensor]: potential for the old and new observations.
```
Comment (shwang) [inline imitation/reward_net.py]: For documentation purposes this is fine, but fun-fact:
`Tuple[tf.Tensor, tf.Tensor]` is actually the return type would make a type-checker pass.

(Tuple is weird; `Sequence[tf.Tensor]` and `Iterable[tf.Tensor]` would have passed okay.)

**Comment by AdamGleave** [comment]
Thanks for the comments Steven, addressed. Holding off on merging since `docstrings` is still open; feel free to merge once that's closed.

**Comment by shwang** [comment]
image

**Pull request merged**